x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -97,12 +107,35 @@ final void startComputeOnDataNodes(
                         nodePermits.putIfAbsent(node, new Semaphore(1));
                     }
                 }
-                pendingShardIds.addAll(targetShards.shards.keySet());
+                pendingShardIds.addAll(order(targetShards));
                 trySendingRequestsForPendingShards(targetShards, computeListener);
             }
         }, listener::onFailure));
     }
 
+    private static List<ShardId> order(TargetShards targetShards) {
+        var ordered = new ArrayList<>(targetShards.shards.keySet());
+        ordered.sort(Comparator.comparingInt(shardId -> nodeOrder(targetShards.getShard(shardId).remainingNodes)));
```

Comment:
Should we move this sort into  the `searchShards` method when we calculate `allocatedNodes`?